src/ast/mod.rs

```diff
@@ -425,11 +384,14 @@ pub enum Expr {
     Identifier(Ident),
     /// Multi-part identifier, e.g. `table_alias.column` or `schema.table.col`
     CompoundIdentifier(Vec<Ident>),
-    /// JSON access (postgres)  eg: data->'tags'
-    JsonAccess {
-        left: Box<Expr>,
-        operator: JsonOperator,
-        right: Box<Expr>,
+    /// Access data nested in a VARIANT on Snowflake, for example `src:customer[0].name`.
+    ///
+    /// See <https://docs.snowflake.com/en/user-guide/querying-semistructured>.
+    VariantAccess {
+        /// The VARIANT being queried.
+        value: Box<Expr>,
+        /// The path to the data inside the VARIANT to extract.
+        path: Vec<VariantPathElem>,
```

Comment:
Thanks for this comment, I appreciate your review! So I did consider that, but my interpretation of this syntax is that `:` is an operator that takes a `VARIANT` expression and a  JSON path expression as a single syntactic element (`VARIANT : JSONPATH`).

Another SQL dialect with the same syntax explicitly says this is the case: 
https://docs.databricks.com/en/sql/language-manual/functions/colonsign.html
https://docs.databricks.com/en/sql/language-manual/sql-ref-json-path-expression.html

(If we do keep this, I should actually rename it back to `JsonAccess` since I do plan to add Databricks as a dialect here in the future...)

Tangentially, there are also some things I'm not happy with in `MapAccess`, particularly `MapAccessSyntax::Dot`:

* `MapAccessSyntax::Dot` overlaps with `CompositeAccess`.
* `MapAccessKey` contains an `Expr` even when using dot syntax, which can only legally contain `Identifier` or `DoubleQuotedString` in that scenario. It would make more sense to use an `Ident` in that case.

So my broader preference here is:

* Use `CompositeAccess` for all non-jsonpath dot accesses.
* Use `MapAccess` for all non-jsonpath bracketed accesses (i.e. eliminate `MapAccessSyntax`).
* Use `JsonAccess` for all jsonpath accesss.

What do you think?

Reply:
> So I did consider that, but my interpretation of this syntax is that : is an operator that takes a VARIANT expression and a JSON path expression as a single syntactic element (VARIANT : JSONPATH).

hmm not sure if that matters a lot from a parsing/ast perpective - in that were not required to represent that literally so we have some leeway to be flexible if it lets us simplify the code/representation?


> So my broader preference here is:

I think a reasonable goal to aim for could be to reuse the same solution for both the map and the variant expressions  - since beyond the leading `:` special case, they are identical? I think this probably means merging both MapAccess and JsonAccess based on your proposal? If that seems reasonable then yeah feel free to make changes to that effect if it means revamping the current MapAccess setup.